test/erlfmt_markdown_SUITE.erl

```diff
@@ -0,0 +1,175 @@
+        find_markdown_filenames_in(RepoRootPath),
+        find_markdown_filenames_in(DocPath)
+    ),
+    ?assertMatch([_One | [_Two | [_Three | _AtLeast]]], Filenames),
+    lists:map(
+        fun (Filename) ->
+            {ok, Content} = file:read_file(Filename),
+            check_markdown(binary_to_list(Content))
+        end,
+        Filenames
+    ).
+
+find_markdown_filenames_in(Path) ->
+    {ok, BaseFilenames} = file:list_dir_all(Path),
+    Filenames =
+        lists:map(fun (Filename) -> filename:join([Path, Filename]) end, BaseFilenames),
+    lists:filter(fun (Filename) -> filename:extension(Filename) == ".md" end, Filenames).
+
+markdown_string(Config) when is_list(Config) ->
+    S =
+        "# Heading\n"
+        "ignore for now\n"
+        "```erlang formatted key\n"
+        "hello(mike, joe, robert)\n"
+        "```\n"
+        "ignore for now\n"
+        "```erlang unformatted\n"
+        "goodbye(mike, joe, robert)\n"
+        "```\n"
+        "The previous unformatted version, should match this formatted version\n"
+        "```erlang formatted\n"
+        "goodbye(mike, joe, robert)\n"
+        "```\n"
+        "This unformatted version, should match the first formatted version with the same key\n"
+        "```erlang unformatted key\n"
+        "hello(mike, joe, robert)\n"
+        "```\n",
+    check_markdown(S).
+
+check_markdown(Content) ->
+    Sections = string:split(Content, "```", all),
+    lists:foldl(fun check_section/2, {text, maps:new()}, Sections).
+
+check_section(_Text, {text, State}) ->
+    {code, State};
+check_section(Text, {code, State}) ->
+    [FirstLine, CurrentCode] = string:split(Text, "\n"),
+    Spec = string:split(FirstLine, " ", all),
+    case Spec of
+        [Erlang | [ToFormat | Keys]] ->
+            case string:prefix(Erlang, "erl") of
+                nomatch ->
+                    {text, State};
+                _ ->
+                    Key = string:join(Keys, " "),
+                    case maps:find(Key, State) of
+                        error ->
+                            case ToFormat of
+                                "formatted" ->
+                                    check_fmt(CurrentCode, CurrentCode),
+                                    {text, maps:put(Key, CurrentCode, State)};
+                                "unformatted" ->
+                                    {text, maps:put(Key, CurrentCode, State)}
+                            end;
+                        {ok, PreviousCode} ->
+                            case ToFormat of
+                                "formatted" ->
+                                    check_fmt(PreviousCode, CurrentCode);
+                                "unformatted" ->
+                                    check_fmt(CurrentCode, PreviousCode)
+                            end,
+                            {text, maps:remove(Key, State)}
+                    end
+            end;
+        _ ->
+            {text, State}
+    end.
```

Comment:
I think this could be clearer if the two phases - gathering chunks to compare and comparison itself were separated.
We could accumulate all the examples into a map like:
```erl
#{Key => {formatted, Formatted} | {unformatted, Unformatted} | {check, Formatted, Unformatted}}
```
 and then at the end traversing the map. This would also allow detecting if keys have typos in them and things are not checked or are accidentally duplicated.

Reply:
Great idea.